### ISSUE TITLE
💡 Implement a plugin system ?

### ISSUE THREAD

**Issue by Nitnelav** [eqasim-org/synpp#66]
WIP : This is just an idea for now...

It would be interesting to be able to imagine a plugin system that allows users to replace a stage with an external one. 
This could be set in the config file in a plugin section.

For exemple : 
stage A depends on stage B. The file structure should look like this :

root
| -- A.py
| -- B.py
| -- config.yml

A.py
```python
def configure(context):
    context.stage("B")

def exec(context):
    B = context.stage("B")
    # ...
```
B.py
```python
def configure(context):
    context.config("bar")

def exec(context):
    foo = context.config("bar")
    # ...
```
config.yml
```yaml
run:
    - A

config:
    bar: "foo"
    
plugins:
    B: /path/to/external/B.py
```

In this exemple, the external B.py is executed instead of the "internal" B.py. It is the responsability of the external B.py to match the data structure expected in the internal B.py outputs.

This would be interesting for big projects like https://github.com/eqasim-org/ile-de-france where somes stages can be solved by various algorithmes or datasources. Down the line this would avoid specifying a lot of "selected" stages with conditionnal switches like in : https://github.com/eqasim-org/ile-de-france/blob/develop/data/hts/selected.py

Can it be done ? How ? How to manage external dependency tree then ? What if dependency C is defined in both trees ?
I don't know. I can work on it if you think it's feasable without breaking everything :)

**Comment by sebhoerl** [comment]
I agree, and I was thinking quite a bit about it and also started a couple of proof-of-concept so to say. I think to have this, we would need to switch to a "toolbox" of stages that perform generic tasks, and then we have a pipeline configuration file like:

```yml
config:
  data_path: /path/to/data
  output_path: /path/to/output

pipeline:
  census_raw:
    stage: data.census.raw
    params:
      rp_path: ${data_path}/insee/RP_2015.dbf
  census_cleaned:
    stage: data.census.cleaned
    input: 
      - census_raw
  my_hts:
    stage: data.entd.cleaned
    params:
       data_path: ${data_path}/entd
  population_base:
    stage: synthesis.population.base
    input:
      - census_cleaned
  population:
    stage: synthesis.statistical_matching.algorithm
    params: 
       matching_fields: [age, sex, socioprofessional_category]
       impute_fields: [number_of_bikes]
    input:
       census: census_cleaned
       hts: my_hts
```

The initial idea was to avoid this kind of huge configuration file. Everything works a bit automatic now in synpp now, because the dependencies are effectively hardcoded. On the other hand, this here is quite handy because the file would be a complete description fo the process that is easy to process and follow and it will be easy to replace things.

We could almost use Snakemake for this purpose, but the problem there is that it is file based. So where I have some named "realizations of a stage", e.g. "my_hts", in Snakemake every stage has a list of output files and input files, and the process is only based on these files. This is a bit less clean that when we would propose here (becasue Snakemake does not detect code changes, which we currently do in synpp).

So basically, thinking big, I would propose a synpp2 to do what you are looking for ;) Or maybe a synpp 1.5 because, I think there is not so much to modify actually to arrive at something like this.

**Comment by Nitnelav** [comment]
All right, nice to know it's something you thought about.
I'd also like to avoid big configuration file and try to stick with the current way of doing things as much as possible.
I'm going to look for a way to go the synpp 1.5 route : changing as little as possible while maintaining backward compatibility.
I have to get familiar with the code first... To be continued...
